woocommerce/woocommerce-gutenberg-products-block
# Convert emitters to promise

While working on #1983 I realized that the event emitters api needs to be implemented using Promises as registered callbacks might need to be promises themselves.  For example, Stripe CC interacts with promises for creating the stripe source.

This also allows the observers to be more asynchronous (even though the implementation here still waits for each observer to resolve).

As a part of this pull:

- I did a bit of restructuring of the `event-emit` directory (moving emit functions outside of the reducer file).
- Added a new `emitEventWithAbort` function that will abort looping through the registered observers on the event if any of them return a non-truthy value. This will help with the handling of validation/errors in the checkout context (or any other future types of events that need similar behaviour).
- Converted `emitEvent` to be a promise.
- updated `CheckoutContextProvider` to correctly implement the emit event changes (`ShippingDataContextProvider` implements `emitEvent`, but no changes were necessary).
- added js unit tests for the `emitEvent` and `emitEventWithAbort` functions.

## To Test

One reason I added the unit tests in this pull is because nothing is currently implementing the event system - so tests verify these functions work as expected. Until we implement observers on the event system there's no user facing tests.

So the testing needed is just doing smoke tests to ensure the Checkout and Shipping contexts don't affect any existing behaviour of the checkout block.
